### PR TITLE
CSP-470 - Added base member classes, validations., auditing services.

### PR DESCRIPTION
**Pull request by LiamWright**
Features implemented:

- Abstracted some common "Member" properties and validations to interfaces and base classes for inheritance
- Added auditing repository

### CODE REVIEW

File: src/SFA.DAS.AANHub.Application/Common/Interfaces/ICreateMemberCommand.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿namespace SFA.DAS.AANHub.Application.Common.Interfaces
+{
+    public interface ICreateMemberCommand
+    {
+        public string? Email { get; set; }
+        public string? Name { get; set; }
+        public DateTime Joined { get; set; }
```
[the remark this answers is not in the document]
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Common/Interfaces/ICreateMemberCommand.cs:7]: I have added an "empty" validation check to `Joined`. I didn't add a unit test as the process to test for an empty datetime was handled by the C# compiler during development anyway.

File: src/SFA.DAS.AANHub.Application/Common/Commands/BaseMemberCommand.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿using SFA.DAS.AANHub.Application.Common.Interfaces;
+
+namespace SFA.DAS.AANHub.Application.Common.Commands
+{
+    public class BaseMemberCommand : ICreateMemberCommand
```
[the remark this answers is not in the document]
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Common/Commands/BaseMemberCommand.cs]: Added.

File: src/SFA.DAS.AANHub.Application/Commands/CreateMember/CreateMemberCommandHandler.cs [hunk 3]
```diff
@@ -13,20 +14,23 @@ public class CreateMemberCommandHandler : IRequestHandler<CreateMemberCommand, C
         private readonly IEmployersWriteRepository _employersWriteRepository;
```
[the remark this answers is not in the document]
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Commands/CreateMember/CreateMemberCommandHandler.cs:14]: That's right, however we created specific tasks to split this functionality into their own separate workflow. To address it as part of this ticket would have been scope creep.

By keeping this handler in, and updating it, the code can be used as a template for subsequent tickets which Ravi and myself will handle and the `CreateMemberHandler/controller` code can be removed after all the other workflows have been completed.

File: src/SFA.DAS.AANHub.Application.UnitTests/Common/Validators/CreateMemberCommandValidatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+using SFA.DAS.AANHub.Application.Commands.CreateMember;
+using SFA.DAS.AANHub.Application.Common.Validators;
+
+namespace SFA.DAS.AANHub.Application.UnitTests.Common.Validators
+{
+    [TestFixture]
+    public class CreateMemberCommandValidatorTests
+    {
+        [TestCase("ThisIsAValidNameLength", true)]
+        [TestCase("ThisIsTooLongOfANameThisIsTooLongOfANameThisIsTooLongOfAName" +
```
[the remark this answers is not in the document]
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application.UnitTests/Common/Validators/CreateMemberCommandValidatorTests.cs]: This is a better way of doing it, and has been implemented.

File: src/SFA.DAS.AANHub.Application/Commands/CreateMember/CreateMemberCommandHandler.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Application/Commands/CreateMember/CreateMemberCommandHandler.cs:14]: @LiamWright can you create a task to remove this so it is not forgotten and link other tasks as `to be done after`.

File: src/SFA.DAS.AANHub.Application/Common/Interfaces/ICreateMemberCommand.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Application/Common/Interfaces/ICreateMemberCommand.cs:7]: What happens to request coming in which is missing that field?

**Review by DevenKShah — APPROVED** [review]
